parrot.rs

```diff
@@ -0,0 +1,73 @@
+use core::sync::atomic::{AtomicUsize, Ordering};
+use kernel::{
+    bindings::msleep,
+    chrdev::Registration,
+    file::{File, Operations, ToUse},
+    io_buffer::IoBufferWriter,
+    prelude::*,
+};
+
+mod frames;
+use frames::*;
+
+module! {
+    type: ParrotSafe,
+    name: b"party_parrot",
+    author: b"John Baublitz (john.m.baublitz@gmail.com)",
+    description: b"Get the party started",
+    license: b"GPL",
+}
+
+const FRAMES: [&str; 10] = [
+    FRAME0, FRAME1, FRAME2, FRAME3, FRAME4, FRAME5, FRAME6, FRAME7, FRAME8, FRAME9,
+];
+
+struct ParrotSafe(Pin<Box<Registration<1>>>);
+
+impl kernel::Module for ParrotSafe {
+    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
+        let mut registration = Registration::new_pinned(name, 0, module)?;
+        registration.as_mut().register::<ParrotOps>()?;
+        Ok(ParrotSafe(registration))
+    }
+}
+
+struct ParrotOps(usize);
+
+impl<'a> Operations for ParrotOps {
+    const TO_USE: ToUse = ToUse {
+        read: true,
+        read_iter: false,
+        write: false,
+        write_iter: false,
+        compat_ioctl: false,
+        fsync: false,
+        ioctl: false,
+        mmap: false,
+        poll: false,
+        seek: false,
+    };
+
+    type Data = Box<AtomicUsize>;
+
+    fn open(_: &Self::OpenData, _: &File) -> Result<Self::Data> {
+        Ok(Box::try_new(AtomicUsize::new(0))?)
+    }
+
+    fn read(data: &AtomicUsize, _: &File, buf: &mut impl IoBufferWriter, _: u64) -> Result<usize> {
+        let frame = FRAMES
+            .get(
+                match data.fetch_update(Ordering::Relaxed, Ordering::Relaxed, |i| {
+                    Some((i + 1) % FRAMES.len())
+                }) {
+                    Ok(i) => i,
+                    Err(i) => i,
+                },
+            )
+            .unwrap_or(&"");
+        buf.write_slice(&frame.as_bytes())?;
```

Comment:
Ideally you should check that `buf` is big enough to hold the frame, otherwise it's going to fail with `EFAULT`.

Reply:
For me to return my own error, I think I need `from_kernel_errno()` to be `pub` not `pub(crate)`. Want me to put up a PR to export that?

Reply:
You can use the exported ones in `kernel::error::code`.

Reply:
Is there a reason why you can't return one of the existing errors?

In any case, I think the right behaviour here would be to call `write_slice` with a buffer capped at the min of `buf.len()` and `frame.as_bytes().len()`.

Reply:
Maybe I should be a little more explicit about what I was going to do. I was going to check that `buf.len() >= frame.len()` and if it wasn't, I was going to return `EINVAL`. I see the codes now so I see why I don't need that exported. Thank you for your suggestion @ojeda! However, is the way I've outlined handling acceptable to you both?

Reply:
I think so. It's your driver, so I guess it's ok for your to require a minimum size to be provided on read.

Reply:
@wedsonaf What would the alternatives be? Only writing as many bytes as are present in `IOBufferWriter`? I think I might prefer that solution to failing.

Reply:
Actually, the ideal solution is for you to not only return a partial buffer, but also to keep track of what was written so far so that in the next read you can pick up from where you left off. The `offset` is meant for this -- if you linearize your frames into a single buffer, you can probably do this easily without even needing the extra atomic state (i.e., you do the modular arithmetic to loop on bytes, not frames).

But for a simple demo of a dancing parrot, I think returning an error or truncating a frame is also acceptable.

Reply:
I've gone ahead and tried the offset solution. It works really nicely and simplifies the code quite a bit!